docs/reference/api-concepts.md

```diff
@@ -0,0 +1,232 @@
+
+See the API documentation for a list of supported content types for each API.
+
+For example:
+
+1. List all of the pods on a cluster in Protobuf format
+
+        GET /api/v1/pods
+        Accept: application/vnd.kubernetes.protobuf
+        ---
+        200 OK
+        Content-Type: application/vnd.kubernetes.protobuf
+        ... binary encoded PodList object
+
+2. Create a pod by sending Protobuf encoded data to the server, but request a response in JSON
+
+        POST /api/v1/namespaces/test/pods
+        Content-Type: application/vnd.kubernetes.protobuf
+        Accept: application/json
+        ... binary encoded Pod object
+        ---
+        200 OK
+        Content-Type: application/json
+        {
+          "kind": "Pod",
+          "apiVersion": "v1",
+          ...
+        }
+
+Not all API resource types will support Protobuf, specifically those defined via Custom Resource Definitions or those that are API extensions. Clients that must work against all resource types should specify multiple content types in their `Accept` header to support fallback to JSON:
+
+```
+Accept: application/vnd.kubernetes.protobuf, application/json
+```
+
+
+### Protobuf encoding
+
+Kubernetes uses an envelope wrapper to encode Protobuf responses, since unlike JSON a Protobuf message is not self-describing (the client cannot determine the type).
+
+The wrapper format is:
+
+```
+A four byte magic number prefix:
+  Bytes 0-3: "k8s\x00" [0x6b, 0x38, 0x73, 0x00]
+
+An encoded Protobuf message with the following IDL:
+  message Unknown {
+    // typeMeta should have the string values for "kind" and "apiVersion" as set on the JSON object
+    optional TypeMeta typeMeta = 1;
+
+    // raw will hold the complete serialized object in protobuf. See the protobuf definitions in the client libraries for a given kind.
+    optional bytes raw = 2;
+
+    // contentEncoding is encoding used for the raw data. Unspecified means no encoding.
+    optional string contentEncoding = 3;
+
+    // contentType is the serialization method used to serialize 'raw'. Unspecified means application/vnd.kubernetes.protobuf and is usually
+    // omitted.
+    optional string contentType = 4;
+  }
+
+  message TypeMeta {
+    // apiVersion is the group/version for this type
+    optional string apiVersion = 1;
+    // kind is the name of the object schema. A protobuf definition should exist for this object.
+    optional string kind = 2;
+  }
+```
+
+Clients that receive a response in `application/vnd.kubernetes.protobuf` that does not match the expected prefix should reject the response since
+future versions may need to alter the serialization format in an incompatible way, and will do so by changing the prefix.
```

Comment:
I'd hope we define a new MIME type instead!

Reply:
Wouldn't help on disk, unfortunately.